### ISSUE TITLE
Operator API Roadmap

### ISSUE THREAD

**Issue by eric-czech** [PrefectHQ/prefect#2690]
Would it be possible to get an explanation of what the ultimate goal is with things like the task implementations in `prefect.tasks.core.operators`?

On one hand, I can easily see the advantage of using it to write small, more demonstrative pipelines that are interchangeable with vanilla Python code, e.g.:

```python
@task
def my_task(x):
    x['v'] *= 2
    return x

with Flow('f') as flow:
    x = dict(v=3)
    y = my_task(x)
    z = y['v'] * 2
state = flow.run()
state.result[z].result
# > 12
```

The fact that this works with or without anything related to Prefect is cool, but I don't understand to what extent real pipelines can and should be written under that assumption.

The Operators, Collections, and Mapping APIs clearly align well with built-in Python functionality though they seem to imply a curiously low level of task granularity for a workflow engine.  Is there a plan in the works to use this very detailed information to optimize DAGs in some way or potentially to make the entire API compatible with regular Python code?  I can't imagine how that's possible without AST macros or something of the like but I thought I'd ask.  

Basically, I'm intrigued by the idea that Prefect didn't commit to a DSL or only an imperative API like many other workflow systems but I can't wrap my head around how practical it is to try to mix DAG definition with vanilla Python code in a real pipeline, or why Prefect encourages such small tasks.  Any thoughts on that would be much appreciated.

**Comment by jlowin** [comment]
Hi @eric-czech, there are a few things I want to dig into here. Most importantly, Prefect doesn't have any opinion on how granular tasks should be, except that they should match whatever your "unit of work" is. When tasks are "too big", it's harder for Prefect to be useful along any dimension - observability, monitoring, checkpointing, retries, etc. On the other hand, making tasks too small can incur some overhead from serialization or contention, as well as making it difficult to map Prefect's fine representation of the DAG onto your coarser conceptual one (though modern tools like Dask mitigate this greatly). Therefore, we probably wouldn't characterize Prefect as "encourag[ing] such small tasks" -- though it absolutely allows them, if they align with your needs.

There's a very simple reason we invested time and research in making the API work the way it does: it makes Prefect more accessible without compromising its power. We know people hate using DSLs and declarative workflows, so we went the other way. However, the closer our syntax gets to "normal" Python, the more any deviation stands out. Therefore, we spend a lot of time examining the places that people encounter friction using our system. For example, this is what resulted in the new [conditional API](https://medium.com/the-prefect-blog/prefect-0-11-0-improved-apis-and-targets-79fef9d90aec) we shipped in 0.11.0.

Therefore, the tangible benefit of features like operators and collections becomes apparent well before we do any form of DAG optimization (which they also enable!). This has a bit to do with Prefect's design philosophy. Since Prefect `Flows` can only contain `Tasks` and `Edges`, the only way to (for example) add two task results together is by introducing another task with appropriate edges. Similarly, the only way to put two task results into a list and pass them to a third task is by introducing an intermediary task and edges. Though this seems straightforward, and is hardly unique to Prefect, the upshot of this is that operations that are trivial in Python (and consequently, to most of our users) are disproportionately complicated in a deferred computation framework. The main reason Prefect includes these operators is to reduce the burden on our users and bring the complexity of those types of operations back to its expected relative level. 

Similarly, mapping is an operation that "feels" simple but is devilishly complex, especially when you contemplate things like reductions and depth-first execution. Most frameworks don't allow it, or force users to resort to hacking around the orchestrator (for example, many Airflow take advantage of the fact that Airflow re-parses DAG files prior to running them in order to generate map dynamics; to us this is tantamount to having workflow logic live in the workflow itself, which is not allowed in Prefect). Therefore, since we know that mapping forms a critical operation for our users, it was paramount to ensure that Prefect had a first-class implementation.

However, just because all this functionality exists doesn't mean you need to incorporate it. That's why we wouldn't say Prefect encourages "such" small tasks -- in truth, the tasks can be as big as you want. Many Prefect customers even run one-task flows to kick off external systems on a schedule. It is true, though, that Prefect wants you to be able to write tasks as small as makes sense for your workflow, because that gives the system more opportunities to help you. So the important thing is that if you DO find yourself with a small unit of work -- like combining two disparate task results into a list for an external API, or multiplying a value by 100 -- Prefect will not force you to compromise or spend disproportionate time.

**Comment by eric-czech** [comment]
Thanks for the thoughtful response @jlowin!  That's very helpful.

For some more context, I work with @hammer (who I think you know) and I'm trying to figure out how to best incorporate Prefect in a bioinformatics workload.  Typical worfklow managers in that space operate on very big or very small units of work via CLI tools, and we'd like to lift much of that out into distributed systems like Dask or Spark.  In a more specific use case, I have a bunch of dataframes based on a variety of data sources and a pipeline that merges/filters/transforms them.  I'd like to have Prefect tasks wrap fairly granular groups of those operations and generally operate on little other than dataframe inputs and outputs.  I've seen some examples of Prefect pipelines that hint at this being a supported use case, but I'd love to know if you're aware of any similar pipelines in the wild or if there are Prefect examples with tasks that operate on Dask/Spark dataframes.  

I don't see any reason it shouldn't work given that running these flows with a LocalExecutor to coordinate the actions taken on distributed Dask/Spark clusters should sidestep a lot of potential serialization issues/inefficiencies.  Any thoughts on potential pitfalls would be awesome though.

p.s.:

> We know people hate using DSLs and declarative workflows, so we went the other way

I'm loving that direction and the conditional API aside, it's been very nice in my prototypes to be able to write pipeline code that reads identically to its python equivalent.  Thanks for sharing the rationale there!

**Comment by jcrist** [comment]
> I've seen some examples of Prefect pipelines that hint at this being a supported use case, but I'd love to know if you're aware of any similar pipelines in the wild or if there are Prefect examples with tasks that operate on Dask/Spark dataframes.

We have plenty of users using prefect to pass pandas dataframes between tasks, or run spark or dask jobs as part of a task. Putting on my dask developer hat, I'm not sure if I'd recommend passing dask dataframes between tasks (serializing futures and reconstituting them on a different worker can be mildly expensive and error prone). Things might work, but I also wouldn't be surprised if they fail. In the next couple months I'm hoping to make the usage of Prefect + Dask more supported and robust, will definitely consider that use case.

---

Given the initial question above appears to be resolved, I'm going to close this for now. Feel free to reopen/open a new issue if you have further questions/issues.
